rfcs/xxxx-collage.md

```diff
@@ -0,0 +1,863 @@
+  measured variance is too high.
+- **Repeated FuseOps**: Some passes (eg `ManifestAlloc`) introduce new calls to primitive function which must be fused
+  and lowered, even though the main work of fusion and lowering has already occurred. We'll need to either
+  retain `FuseOps`, or ensure `CollageFuseOps` retains the efficiency and handling of `FuseOps` when there's no
+  toolchain ambiguity.
+- **Explainability**: It's easy to show the user the final partitioning and estimated times for each kernel, but harder
+  to show why that partitioning won out over all others during search.
+- **Does not subsume `partition_for_<toolchain>`**: We don't have any plan to deprecate the existing patterns of each
+  BYOC integration supplying a `partiion_for_<toolchain>` function. If the user has a specific toolchain in mind then
+  making the partition explicit enjoys both faster compilation and can incorporate global optimization passes which
+  Collage cannot currently account for (eg enforcing a particular layout).
+
+## Sub-projects
+
+These items need more design and can be run as 'sub-projects'.
+
+### Robust candidate kernel latency measurement
+
+Collage requires an implementation of a `CostEstimator`:
+
+```
+class CostEstimator {
+ public:
+  /*!
+   * \brief Return the estimated cost (possibly after many many minutes of training time) of
+   * running function using target.
+   */
+  virtual Cost Estimate(const Function& function, const Target& target) const;
+}
+```
+
+The 'v2' prototype has implemented this with an in-memory cache and a small Python driver which defers to
+TVM's `tvm.runtime.vm.VirtualMachine`s `benchmark` helper. The following needs to be designed and implemented:
+
+- Compilation should be in units of `IRModule` rather than `Function` so that, in the future, additional global
+  definitions (such as for weights) can be conveyed to the toolchain.
+- The recent MetaSchedule work has provided `BuilderInput` (`include/tvm/meta_schedule/builder.h`),
+  `RunnerInput` (`include/tvm/meta_schedule/runner.h`) and `Database` (`include/tvm/meta_schedule/database.h`)
+  interfaces. The latter is for `TuningRecord`s of `Workload`s. It looks like these interfaces can support the
+  measurement of Collage `CandidateKernel`s with minor changes.
+- (Internal to OctoML) We need an implementation connecting to the internal OctoML kernel tuning workflow and production
+  cache. Ideally this would be the same implementation as for the MetaSchedule system.
+- Collage converts measured 50th %ile latencies to costs in seconds. We may need to consider taking a slightly higher
+  %ile to be more robust against variance on small kernels. We need to validate the estimated variance reflects true
+  variance.
+- For TVM-native targets, we would like the `Estimate` call to perform any TVM tuning required for a novel candidate
+  kernel.
+
+### Easier Library Integration
```

Comment:
Though moved into Appendix B I 'd still like to keep a placeholder for this issue since it motivates how I'm proposing to support "Composite" functions fairly generically. I'll be happy to delete the 'One possibility is...' section once you've got something to replace it.